### ISSUE TITLE
Move go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/internal/semconvutil into otelhttp

### ISSUE THREAD

**Issue by els0r** [open-telemetry/opentelemetry-go-contrib#4580]
### Problem Statement

The `semconvutil` package follows semantic conventions for span attributes set during a round-trip of a client-request.

Exposing the utility functions from `semconvutil` would allow other `RoundTripper`s to use them, leading to consistent, semconv-aligned telemetry.

In my specific use case, I'm providing a logging transport and would like to use the same attribute keys as the `otelhttp` transport does.

### Proposed Solution

Move `semconvutil` from `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/internal/semconvutil` to `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/semconvutil`

#### Alternatives

Copy-pasting the internals of `semconvutil` to internal instrumentation library. Which isn't really maintainable.

Or keep the current implementation, which doesn't align with the semantic conventions. Example: 

#### Prior Art

Not aware of any.

### Additional Context

We have some tools which run a low frequency of client requests and would like to correlate log messages in our system with the corresponding spans we export.

**Comment by MrAlias** [comment]
I don't think we want to expose this package without a serious review. It is not designed with stability in mind and is likely going to have major breaking changes as the semantic conventions evolve.

**Comment by els0r** [comment]
Thanks for the context on https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/internal/shared/semconvutil

Even if the semantic conventions evolve, it would be nice to stay consistent with the ones that are currently in use by the instrumentation libraries in question.

What would you recommend as a way to go forward? What I can think of right now is to 

* generate `httpconv.go` from the template and store it inside our own instrumentation librarie's `internal` folder
* monitor for changes of that file (as in: newer versions of `semconv` being referenced)
* rinse & repeat

Does that make sense or do you see another route to re-use as much code as possible?

**Comment by pellared** [comment]
> Does that make sense or do you see another route to re-use as much code as possible?

It makes sense. This is what I suggested a few times.

**Comment by StarpTech** [comment]
Hi, what's the status? I want to implement my own transport instrumentation but can't use the same semconvutil because the package is internal.

**Comment by dmathieu** [comment]
This hasn't moved yet. The stability of otelhttp, and semconvutil is still not there.

**Comment by StarpTech** [comment]
I got it, but we already rely on it in the instrumentations. What's the difference of consuming them directly from using them indirectly? In both cases, a breaking change could break my system. Therefore e.g. `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` is still under v1. An option could be to export the module through the instrumentation libraries. That would help me to ship customization's.

**Comment by dmathieu** [comment]
There's a difference between providing backwards compatibility with the provided intrumentation (the attributes remain compatible), and providing a backwards compatible API. The latter is a much bigger constraint.

**Comment by StarpTech** [comment]
There are no API guarantees below v1. Several breaking changes has been made since. https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4707#discussion_r1484515166 I don't want to start a general discussion about backwards compatibility. I just the need that people want to rely on the semconvutil to not reinvent the wheel. I see no reason to export them with the same guarantees as the modules which rely on it.

**Comment by pellared** [comment]
I think we could consider exporting (exposing) helper functions in `otelhttp` that create spans and metrics for incoming and outgoing requests. This could would allow libraries that work on top of `net/http` to "natively" instrument the package or make it easier to create instrumentation libraries outside of https://github.com/open-telemetry/opentelemetry-go-contrib.

I think it will not have that much issues like we did when  `semconvutil` was in https://github.com/open-telemetry/opentelemetry-go. We would just need to make sure to properly design the API to make it usable and backwards-compatible.

**Comment by tina-junold** [comment]
I'm having the same problem. For me, with the "tempoary" deprecation of otelecho, it was the point to create an own implementation. But this is nearly impossible unless you have access to semconvutil without reinvent the wheel for attributes and status.

```go
oteltrace.WithAttributes(semconvutil.HTTPServerRequest(service, request)...),
//[...]
span.SetStatus(semconvutil.HTTPServerStatus(status))
```

In my case, I created a middlware that handle metrics, logs & traces in one place which has some benefits (for me). but the only way was to "copy" the "semconvutil" to my code whitch is not a viable solution.

**Comment by StarpTech** [comment]
@pellared main reason is the rapid evolution of semconv attributes in the middleware. We distribute software and rely on the exact attribute naming. We can't just upgrade and don't want to when naming changes or new attributes are added.

**Comment by tina-junold** [comment]
Any news on this?

**Comment by flc1125** [comment]
I am wondering if we can consider referencing the approach of [Google APIs](https://github.com/googleapis/googleapis/tree/master/google/iam), using concepts like beta versions to manage the instability of `semconv`.

**Comment by dmathieu** [comment]
We are not going to move `semconvutil` to be public. That package isn't properly thought-out to be made public and stable.
However, we're looking into introducing a new package that would do something similar, but thought to be public from the ground up.

Feel free to review and provide constructive feedback on https://github.com/open-telemetry/opentelemetry-go-contrib/pull/6859

**Comment by tina-junold** [comment]
@dmathieu So the topic is dead? since #6859 is closed.

**Comment by dmathieu** [comment]
The idea of a library that would allow instrumenting any http interface isn't necessarily dead. But it's definitely not a priority.

**Comment by tina-junold** [comment]
@dmathieu thank you for the response. i see definitely a use case for such a library, IMHO the otelecho is not yet compatible with echo v5. another thing would be custom trace handling in a middleware...

**Comment by dmathieu** [comment]
See https://github.com/open-telemetry/opentelemetry-go-contrib/pull/8457 and https://github.com/labstack/echo-contrib/pull/141 regarding echo.

**Comment by tina-junold** [comment]
am i wrong or did they copy the semconvutil to the contrib lib?

**Comment by dmathieu** [comment]
The copied the generated internal semconv package, yes.

**Comment by aldas** [comment]
Hi,  Echo maintainer here. We created new middleware for Echo v5 https://github.com/labstack/echo-opentelemetry  

At first the plan was to just plain copy `semconv` package but I decide to rewrite.  It felt that the current `semconv` API is little bit too complex for middleware serving HTTP request/responses based on Go standard library `http` package.  
Also the documentation in `semconv` package lacked and we all know that reading the manual (spec) is for the weak. It is easier to read doccomments etc inside code than to open spec website and try to find relevant information there.

Populating a struct from the request/response and exposing lightweight helper methods for getting span starting/ending and  metric attributes would provide greater flexibility for framework or library developers, while reducing the responsibilities handled directly by OpenTelemetry.  Current `semconv` is quite rigid/opinionated in that side.

If Open Telemetry is interested I can contribute/give this https://github.com/labstack/echo-opentelemetry/blob/main/extrator.go part to the project and modify it to pass the quality/naming etc standards. 

The usage is visible in middleware:
https://github.com/labstack/echo-opentelemetry/blob/main/otel.go#L171
